File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/containers/NavigationContainerView.kt
package com.crescentflare.piratesgame.components.containers

import android.annotation.TargetApi
import android.content.Context
import android.content.res.Resources
import android.graphics.Rect
import android.os.Build
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.piratesgame.R
import com.crescentflare.piratesgame.components.utility.NavigationBarComponent
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.piratesgame.infrastructure.inflator.Inflators
import java.lang.ref.WeakReference

/**
 * Container view: contains the layout framework of a page, a content view and navigation bar components with optional translucency
 */
class NavigationContainerView: ViewGroup, AppEventObserver {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return NavigationContainerView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is NavigationContainerView) {
                    // Set content
                    val contentItem = Inflators.viewlet.attributesForNestedInflatable(attributes["content"])
                    val recycling = mapUtil.optionalBoolean(attributes, "recycling", false)
                    if (recycling && Inflators.viewlet.canRecycle(obj.contentView, contentItem)) {
                        val contentView = obj.contentView
                        if (contentView != null && contentItem != null) {
                            Inflators.viewlet.inflateOn(contentView, contentItem, null, binder)
                            ViewletUtil.applyLayoutAttributes(mapUtil, contentView, contentItem)
                            ViewletUtil.bindRef(mapUtil, contentView, contentItem, binder)
                        }
                    } else {
                        // First empty content
                        obj.contentView = null

                        // Set content view
                        val result = Inflators.viewlet.inflate(obj.context, contentItem, obj, binder)
                        if (result is View && contentItem != null) {
                            obj.contentView = result
                            ViewletUtil.applyLayoutAttributes(mapUtil, result, contentItem)
                            ViewletUtil.bindRef(mapUtil, result, contentItem, binder)
                        }
                    }

                    // Set top bar
                    val topBarItem = Inflators.viewlet.attributesForNestedInflatable(attributes["topBar"])
                    if (recycling && Inflators.viewlet.canRecycle(obj.topBarView, topBarItem)) {
                        val topBarView = obj.topBarView
                        if (topBarView != null && topBarItem != null) {
                            Inflators.viewlet.inflateOn(topBarView, topBarItem, null, binder)
                            ViewletUtil.applyLayoutAttributes(mapUtil, topBarView, topBarItem)
                            ViewletUtil.bindRef(mapUtil, topBarView, topBarItem, binder)
                        }
                    } else {
                        // First empty top bar
                        obj.topBarView = null

                        // Set top bar view
                        val result = Inflators.viewlet.inflate(obj.context, topBarItem, obj, binder)
                        if (result is View && topBarItem != null) {
                            obj.topBarView = result
                            ViewletUtil.applyLayoutAttributes(mapUtil, result, topBarItem)
                            ViewletUtil.bindRef(mapUtil, result, topBarItem, binder)
                        }
                    }

                    // Set bottom bar
                    val bottomBarItem = Inflators.viewlet.attributesForNestedInflatable(attributes["bottomBar"])
                    if (recycling && Inflators.viewlet.canRecycle(obj.bottomBarView, bottomBarItem)) {
                        val bottomBarView = obj.bottomBarView
                        if (bottomBarView != null && bottomBarItem != null) {
                            Inflators.viewlet.inflateOn(bottomBarView, bottomBarItem, null, binder)
                            ViewletUtil.applyLayoutAttributes(mapUtil, bottomBarView, bottomBarItem)
                            ViewletUtil.bindRef(mapUtil, bottomBarView, bottomBarItem, binder)
                        }
                    } else {
                        // First empty bottom bar
                        obj.bottomBarView = null

                        // Set bottom bar view
                        val result = Inflators.viewlet.inflate(obj.context, bottomBarItem, obj, binder)
                        if (result is View && bottomBarItem != null) {
                            obj.bottomBarView = result
                            ViewletUtil.applyLayoutAttributes(mapUtil, result, bottomBarItem)
                            ViewletUtil.bindRef(mapUtil, result, bottomBarItem, binder)
                        }
                    }

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is NavigationContainerView
            }
        }
    }


    // --
    // Members
    // --

    private var eventObserverReference : WeakReference<AppEventObserver>? = null
    private val actionBarHeight: Int
    private var solidTopBar = false
    private var solidBottomBar = false


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        actionBarHeight = getActionBarHeight()
    }


    // --
    // Configurable values
    // --

    var eventObserver: AppEventObserver?
        get() = eventObserverReference?.get()
        set(newValue) {
            eventObserverReference = if (newValue != null) {
                WeakReference(newValue)
            } else {
                null
            }
        }

    var contentView: View? = null
        set(contentView) {
            if (field != null) {
                removeView(field)
            }
            field = contentView
            if (field != null) {
                addView(field, 0)
            }
        }

    var topBarView: View? = null
        set(topBarView) {
            if (field != null) {
                removeView(field)
            }
            field = topBarView
            if (field != null) {
                addView(field)
            }
            solidTopBar = field != null && !((field as? NavigationBarComponent)?.translucent == true)
        }

    var bottomBarView: View? = null
        set(bottomBarView) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) { // Ignore on older Android versions
                if (field != null) {
                    removeView(field)
                }
                field = bottomBarView
                if (field != null) {
                    addView(field)
                }
                solidBottomBar = field != null && !((field as? NavigationBarComponent)?.translucent == true)
            }
        }


    // --
    // Interaction
    // --

    override fun observedEvent(event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender)
    }


    // --
    // Handle insets
    // --

    val safeInsets: Rect
        get() = Rect(0, if (!solidTopBar) transparentStatusBarHeight + actionBarHeight else 0, 0, if (!solidBottomBar) transparentNavigationBarHeight else 0)

    val transparentStatusBarHeight: Int
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && Resources.getSystem().displayMetrics.widthPixels < Resources.getSystem().displayMetrics.heightPixels) rootWindowInsets.stableInsetTop else 0

    private val transparentNavigationBarHeight: Int
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && Resources.getSystem().displayMetrics.widthPixels < Resources.getSystem().displayMetrics.heightPixels) rootWindowInsets.stableInsetBottom else 0


    // --
    // Custom layout
    // --

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = MeasureSpec.getSize(widthMeasureSpec)
        val height = MeasureSpec.getSize(heightMeasureSpec)
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        if (widthMode == MeasureSpec.EXACTLY && heightMode == MeasureSpec.EXACTLY) {
            val totalTopBarHeight = actionBarHeight + transparentStatusBarHeight
            val bottomBarHeight = transparentNavigationBarHeight
            val topContentInset = if (solidTopBar) totalTopBarHeight else 0
            val bottomContentInset = if (solidBottomBar) bottomBarHeight else 0
            (topBarView as? NavigationBarComponent)?.statusBarInset = transparentStatusBarHeight
            topBarView?.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(totalTopBarHeight, MeasureSpec.EXACTLY))
            bottomBarView?.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(bottomBarHeight, MeasureSpec.EXACTLY))
            contentView?.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(height - topContentInset - bottomContentInset, MeasureSpec.EXACTLY))
            setMeasuredDimension(width, height)
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        }
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val totalTopBarHeight = actionBarHeight + transparentStatusBarHeight
        val bottomBarHeight = transparentNavigationBarHeight
        val topContentInset = if (solidTopBar) totalTopBarHeight else 0
        val bottomContentInset = if (solidBottomBar) bottomBarHeight else 0
        (topBarView as? NavigationBarComponent)?.statusBarInset = transparentStatusBarHeight
        topBarView?.layout(0, 0, right - left, totalTopBarHeight)
        bottomBarView?.layout(0, bottom - top - bottomBarHeight, right - left, bottom - top)
        contentView?.layout(0, topContentInset, right - left, bottom - top - bottomContentInset)
    }


    // --
    // Helper
    // --

    private fun getActionBarHeight(): Int {
        val typedValue = TypedValue()
        return if (context.theme.resolveAttribute(R.attr.actionBarSize, typedValue, true)) TypedValue.complexToDimensionPixelSize(typedValue.data, Resources.getSystem().displayMetrics) else 0
    }

}

File: pirates-ios/PiratesGame/Components/Containers/NavigationContainerView.swift
//
//  NavigationContainerView.swift
//  Container view: contains the layout framework of a page, a content view and navigation bar components with optional translucency
//

import UIKit
import UniLayout
import JsonInflator

enum NavigationContainerViewScrollPaddingType: String {
    
    case none = "none"
    case topAndBottom = "topAndBottom"
    case statusAndBottom = "statusAndBottom"
    
}

class NavigationContainerView: UIView, AppEventObserver {
    
    // --
    // MARK: Members
    // --
    
    weak var viewController: NavigationViewController?
    
    
    // --
    // MARK: Viewlet integration
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return NavigationContainerView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let navigationContainer = object as? NavigationContainerView {
                // Update content view
                let contentItem = Inflators.viewlet.attributesForNestedInflatable(attributes["content"])
                let recycling = convUtil.asBool(value: attributes["recycling"]) ?? false
                if recycling && Inflators.viewlet.canRecycle(object: navigationContainer.contentView, attributes: contentItem) {
                    if let item = contentItem {
                        if let view = navigationContainer.contentView {
                            Inflators.viewlet.inflate(onObject: view, attributes: item, binder: binder)
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                } else {
                    // First empty content
                    navigationContainer.contentView = nil
                    
                    // Set content item
                    if let item = contentItem {
                        if let view = Inflators.viewlet.inflate(attributes: item, parent: navigationContainer, binder: binder) as? UIView {
                            navigationContainer.contentView = view
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                }
                
                // Update top bar
                let topBarItem = Inflators.viewlet.attributesForNestedInflatable(attributes["topBar"])
                if recycling && Inflators.viewlet.canRecycle(object: navigationContainer.topBarView, attributes: topBarItem) {
                    if let item = topBarItem {
                        if let view = navigationContainer.topBarView {
                            Inflators.viewlet.inflate(onObject: view, attributes: item, binder: binder)
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                } else {
                    // First empty content
                    navigationContainer.topBarView = nil
                    
                    // Set bar item
                    if let item = topBarItem {
                        if let view = Inflators.viewlet.inflate(attributes: item, parent: navigationContainer, binder: binder) as? UIView {
                            navigationContainer.topBarView = view
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                }
                
                // Update bottom bar
                let bottomBarItem = Inflators.viewlet.attributesForNestedInflatable(attributes["bottomBar"])
                if recycling && Inflators.viewlet.canRecycle(object: navigationContainer.bottomBarView, attributes: bottomBarItem) {
                    if let item = bottomBarItem {
                        if let view = navigationContainer.bottomBarView {
                            Inflators.viewlet.inflate(onObject: view, attributes: item, binder: binder)
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                } else {
                    // First empty content
                    navigationContainer.bottomBarView = nil
                    
                    // Set bar item
                    if let item = bottomBarItem {
                        if let view = Inflators.viewlet.inflate(attributes: item, parent: navigationContainer, binder: binder) as? UIView {
                            navigationContainer.bottomBarView = view
                            ViewletUtil.applyLayoutAttributes(convUtil: convUtil, view: view, attributes: item)
                            ViewletUtil.bindRef(view: view, attributes: item, binder: binder)
                        }
                    }
                }
                
                // Linked scroll container
                let linkedScrollContainer = (binder as? InflatorDictBinder)?.findByReference(convUtil.asString(value: attributes["linkedScrollContainer"]) ?? "")
                navigationContainer.linkedScrollContainer = linkedScrollContainer as? ScrollContainerView
                navigationContainer.automaticScrollPadding = NavigationContainerViewScrollPaddingType(rawValue: convUtil.asString(value: attributes["automaticScrollPadding"]) ?? "") ?? .none
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: navigationContainer, attributes: attributes)
                
                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    navigationContainer.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is NavigationContainerView
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    
    // --
    // MARK: Configurable values
    // --
    
    weak var eventObserver: AppEventObserver?
    
    var contentView: UIView? {
        didSet {
            oldValue?.removeFromSuperview()
            if let contentView = contentView {
                insertSubview(contentView, at: 0)
            }
        }
    }
    
    var topBarView: UIView? {
        didSet {
            oldValue?.removeFromSuperview()
            if let topBarView = topBarView {
                addSubview(topBarView)
            }
            updateLinkedScrollPadding()
        }
    }
    
    var bottomBarView: UIView? {
        didSet {
            oldValue?.removeFromSuperview()
            if let bottomBarView = bottomBarView {
                addSubview(bottomBarView)
            }
            updateLinkedScrollPadding()
        }
    }
    
    weak var linkedScrollContainer: ScrollContainerView? {
        didSet {
            updateLinkedScrollPadding()
        }
    }
    
    var automaticScrollPadding: NavigationContainerViewScrollPaddingType = .none {
        didSet {
            updateLinkedScrollPadding()
        }
    }
    
    
    // --
    // MARK: Interaction
    // --
    
    func observedEvent(_ event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender: sender)
    }
    
    
    // --
    // MARK: Handle insets
    // --
    
    private func updateLinkedScrollPadding() {
        if let linkedScrollContainer = linkedScrollContainer {
            if automaticScrollPadding != .none && (topBarView as? NavigationBarComponent)?.isTranslucent ?? false {
                linkedScrollContainer.extraTopInset = automaticScrollPadding == .statusAndBottom ? statusBarHeight() : topBarHeight()
            } else {
                linkedScrollContainer.extraTopInset = 0
            }
            if automaticScrollPadding != .none && (bottomBarView as? NavigationBarComponent)?.isTranslucent ?? false {
                linkedScrollContainer.extraBottomInset = bottomBarHeight()
            } else {
                linkedScrollContainer.extraBottomInset = 0
            }
        }
    }
    
    
    // --
    // MARK: Custom layout
    // --
    
    override func layoutSubviews() {
        var topInset: CGFloat = 0
        var bottomInset: CGFloat = 0
        updateLinkedScrollPadding()
        if let topBarView = topBarView {
            let barHeight = topBarHeight()
            (topBarView as? NavigationBarComponent)?.statusBarInset = min(UIApplication.shared.statusBarFrame.width, UIApplication.shared.statusBarFrame.height)
            UniLayout.setFrame(view: topBarView, frame: CGRect(x: 0, y: 0, width: bounds.width, height: barHeight))
            topInset = (topBarView as? NavigationBarComponent)?.isTranslucent ?? false ? 0 : barHeight
        }
        if let bottomBarView = bottomBarView {
            let barHeight = bottomBarHeight()
            UniLayout.setFrame(view: bottomBarView, frame: CGRect(x: 0, y: bounds.height - barHeight, width: bounds.width, height: barHeight))
            bottomInset = (bottomBarView as? NavigationBarComponent)?.isTranslucent ?? false ? 0 : barHeight
        }
        if let contentView = contentView {
            UniLayout.setFrame(view: contentView, frame: CGRect(x: 0, y: topInset, width: bounds.width, height: bounds.height - topInset - bottomInset))
        }
    }
    
    
    // --
    // MARK: Helper
    // --
    
    func statusBarHeight() -> CGFloat {
        return min(UIApplication.shared.statusBarFrame.width, UIApplication.shared.statusBarFrame.height)
    }
    
    func topBarHeight() -> CGFloat {
        var height: CGFloat = 0
        height += min(UIApplication.shared.statusBarFrame.width, UIApplication.shared.statusBarFrame.height)
        if let navigationController = viewController?.navigationController {
            height = navigationController.navigationBar.frame.origin.y
            height += navigationController.navigationBar.frame.height
        }
        return height
    }
    
    func bottomBarHeight() -> CGFloat {
        if #available(iOS 11.0, *) {
            return safeAreaInsets.bottom
        }
        return 0
    }
    
}
